### ISSUE TITLE
warn if both `tensorboard` and `tb-nightly` pip packages are installed

### ISSUE THREAD

**Issue by nfelt** [tensorflow/tensorboard#1907]
This is a common source of inscrutable errors, because pip will happily install both pip packages even though they ship the tensorboard python package [1], and in doing so the second one will overwrite any files from the first one: https://github.com/pypa/pip/issues/4625.  This means you can easily end up with a garbled frankentensorboard package (and it gets even worse if you uninstall one of them but not both of them, since then you get a frankentensorboard with a lot of vital organs removed).

Until pip fixes this mess, it's a fact of life, but we can at least try to warn users about this pitfall ourselves.

[1] It's colloquial to call both of these packages, so I'm prefixing to avoid confusion.  See also https://github.com/pypa/packaging.python.org/issues/106

**Comment by nfelt** [comment]
For an example of this issue happening, see https://github.com/tensorflow/tensorboard/issues/1862

**Comment by nfelt** [comment]
We could warn about this at runtime if we used the `pkg_resources` module from setuptools to check active distributions (aka pip packages).  I think this would be something like `pkg_resources.find(pkg_resources.Requirement("tensorboard"))`.

Even better would be to do this check at installation time.  I think this may be possible via the `cmdclass` parameter to `setup()` by passing in a custom install command, see: 
https://stackoverflow.com/questions/20288711/post-install-script-with-python-setuptools
https://blog.niteo.co/setuptools-run-custom-code-in-setup-py/

Either way we'd want to check reasonably thoroughly to make sure we don't have false positives.  It's also TBD whether to just log a warning or whether to fail entirely.  I'm inclined towards the latter but it might be good to start out with the former just in case there are any false positive installation scenarios out there.  (We could always offer some kind of escape hatch if you get a false positive, like setting an environment variable that disables the check.)

**Comment by nfelt** [comment]
Another example of this happening: https://github.com/tensorflow/tensorboard/issues/2092

**Comment by wchargin** [comment]
Note that you get in this state if you upgrade to the TensorFlow 2.0
alpha using the recommended incantation from the [Dev Summit talk]:

```
$ virtualenv -q ./ve
$ . ./ve/bin/activate
(ve) $ pip install tensorflow==1.13.1 >/dev/null 2>&1
(ve) $ pip install -U --pre tensorflow >/dev/null 2>&1
(ve) $ pip freeze | grep -e tensor -e tf- -e tb-
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
tb-nightly==1.14.0a20190301
tensorboard==1.13.1
tensorflow==2.0.0a0
tensorflow-estimator==1.13.0
tf-estimator-nightly==1.14.0.dev2019030115
```

[Dev Summit talk]: https://youtu.be/k5c-vg4rjBw?t=102

**Comment by nfelt** [comment]
At some point in the lead-up to the 2.0 final release we'll want to release a TensorBoard 2.0 alpha under the `tensorboard` package name, which will resolve that, but it's a good point that at least for now it's easy for users to arrive in this state by following our own instructions.

It would really be ideal if we could auto-uninstall `tensorboard` when installing `tb-nightly` and vice versa, but I think even if we did that as an install command (which is already a little risky/surprising) it might not work if the files are already copied over.  The only way I know of that might reliably work (at least for sufficiently recent versions of pip) is to publish `0.0` versions of both `tensorboard` and `tb-nightly` that contain 0 actual python files, and then have each real package depend on the other with a pin to the `0.0` version, but this is an ugly hack.

**Comment by janosh** [comment]
> ...but it's a good point that at least for now it's easy for users to arrive in this state by following our own instructions.

Exactly. This happened to me. Didn't realize for quite some time that I had an old TB installation lying around.

**Comment by wchargin** [comment]
Quick update: TensorBoard itself doesn’t warn you about this failure
mode, but our self-diagnosis script does:
<https://raw.githubusercontent.com/tensorflow/tensorboard/master/tensorboard/tools/diagnose_tensorboard.py>

If you run it, you’ll get a message like:

> ```
> WARNING: conflicting installations: ['tb-nightly', 'tensorboard']
> ```
>
> ### Suggestion: Fix conflicting installations
>
> Conflicting package installations found. Depending on the order of
> installations and uninstallations, behavior may be undefined. Please
> uninstall ALL versions of TensorFlow and TensorBoard, then reinstall
> ONLY the desired version of TensorFlow, which will transitively pull
> in the proper version of TensorBoard. (If you use TensorBoard without
> TensorFlow, just reinstall the appropriate version of TensorBoard
> directly.)
>
> Namely:
>
>     pip uninstall tb-nightly tensorboard
>     pip install tensorflow  # or `tensorflow-gpu`, or `tf-nightly`, ...

It would still be nice to do this at runtime, too, as described in this
issue.
